design/secrets_decoding_design.md

```diff
@@ -0,0 +1,207 @@
+```
+conjur.org/conjur-secrets.groupname: |
+  - user: policy/path/base64_user     # valid annotation with content-type
+    content-type: base64
+  - policy/path/base64_password:      # this is also valid when no alias is needed
+      content-type: base64
+```
+
+P2F already leverages Kubernetes' Downward API and volumeMounts to pass these annotations into the Secrets Provider container. This will make it possible to determine which secrets should be decoded before pushing them to the Kubernetes secret file. 
+
+## Design
+The goal of this design is to make usage of this feature as consistent as possible across different configurations. Due to the nature of configuring Conjur secrets in K8s secrets mode versus P2F mode, it makes the most sense to add the configuration in-place where Conjur secrets are already defined.
+* Kubernetes Secrets mode - add content-type to the existing K8s secrets manifest
+* P2F mode - add content-type to the existing pod annotations
+
+In both modes the solution will need to be backwards compatible with existing annotation formats. 
+
+The logic and structs should added for this feature should be shared between both modes when possible. For example, P2F mode uses [SecretSpec struct](https://github.com/cyberark/secrets-provider-for-k8s/tree/main/pkg/secrets/common/secret_spec.go) which could be updated with a ContentType field. Ideally both YAML formats can be unmarshaled into a common struct since they will contain the same data:
+* Alias
+* Path
+* ContentType
+  * Allowed values = ['text', 'base64']
+  * Default = 'text'
+
+### Kubernetes Secrets Mode
+In order to keep Conjur secrets configuration from creeping into other locations, we will add the content-type under the existing `conjur-map` annotation of the secrets manifest. This requires adding an additional supported multiline format:
+```
+conjur-map: |-
+  <alias_value>: <path_value>                 # supported
+  <alias_value>:                              # to be added
+    path: <path_value>
+    content-type: <content_type_value>    
+```
+
+### P2F Mode
+For P2F, we can update the existing `conjur.org/conjur-secrets.{secret-group}` annotations with the following allowed formats:
+```
+conjur.org/conjur-secrets.groupname: |
+  - <alias_value>: <path_value>               # supported
+  - <path_value>                              # supported
+  - <alias_value>: <path_value>               # to be added
+    content-type: <content_type_value>
+  - <path_value>:                             # to be added
+      content-type: <content_type_value>
+```
+
+### Backwards compatibility
+This enhancement will not modify the default behavior of Secrets Provider. It will continue to assume that secrets should be delivered "as-is" unless it specifically finds a content-type annotation which indicates that the secret value should be decoded.
+
+## Test Plan
```

Comment:
Is it a correct assumption that if someone mistakenly puts the same secret twice in the annotation and gives them different types, the one defined later in the file would be the type used? Do we need any tests around that?

Reply:
That would likely be the case if they are exactly the same alias/path and in the same manifest.

Although we should support the same Conjur secret with multiple content-types as long as it is in different secret groups (P2F) or under a different kubernetes secret. 

For example:
```
conjur.org/conjur-secrets.decoded-group: |
  - mysecret: /path/to/secret
    content-type: base64
conjur.org/conjur-secrets.encoded-group: |
  - mysecret: /path/to/secret
    content-type: text
```
Both definitions can exist at the same time but the value for mysecret in `decoded-group` will be decoded while the value for mysecret in `encoded-group` will be provided as its original base64 value.